### PR TITLE
Add general object detection estimator for PyTorch

### LINKED ISSUE #1196: General estimator for object detectors in PyTorch
We would like to investigate if it would be possible to implement a general estimator for object detection models in PyTorch called `PyTorchObjectDetector`, similar to `PyTorchClassifier` for general classification models.

`PyTorchObjectDetector` should be able to support various models like:
- torchvision: https://pytorch.org/vision/stable/models.html#object-detection-instance-segmentation-and-person-keypoint-detection
- Detectron2: https://github.com/facebookresearch/detectron2/blob/master/MODEL_ZOO.md#coco-object-detection-baselines
- GoNet: https://github.com/amoudgl/pygoturn
- etc.

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request adds a general, model-independent, object detection estimator for PyTorch models in `art.estimators.object_detection.PyTorchObjectDetector`.

Fixes #1196

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: art/estimators/object_detection/python_object_detector.py [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+            else:
+                for img in inputs_t:
+                    gradients = img.grad.copy()
+                    grad_list.append(gradients)
+
+        if isinstance(x, np.ndarray):
+            grads = np.stack(grad_list, axis=0)
+        else:
+            grads = torch.stack(grad_list, dim=0)
+        grads = np.transpose(grads, (0, 2, 3, 1))
```
Comment (GiulioZizzo) [inline art/estimators/object_detection/python_object_detector.py:313]: if grads could be a tensor should we have a torch.transpose option as well?

**Review by GiulioZizzo — COMMENTED** [review]
Hi Beat, thank you for the review and I have left a few comments!

File: art/estimators/object_detection/python_object_detector.py [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+
+                  - boxes (FloatTensor[N, 4]): the boxes in [x1, y1, x2, y2] format, with 0 <= x1 < x2 <= W and
+                                               0 <= y1 < y2 <= H.
+                  - labels (Int64Tensor[N]): the labels for each image
+        :return: Loss gradients of the same shape as `x`.
+        """
+        import torch  # lgtm [py/repeated-import]
+        import torchvision  # lgtm [py/repeated-import]
+
+        self._model.train()
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/estimators/object_detection/python_object_detector.py:177]: Yes, I think it should be correct because the loss terms are only calculated/available in train mode.

File: art/estimators/object_detection/python_object_detector.py [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    def device(self) -> "torch.device":
+        """
+        Get current used device.
+
+        :return: Current used device.
+        """
+        return self._device
+
+    def _get_losses(
+        self, x: np.ndarray, y: Union[List[Dict[str, np.ndarray]], List[Dict[str, "torch.Tensor"]]]
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/estimators/object_detection/python_object_detector.py:160]: For now we don't support `torch.Tensor` for `x` and checked in line 182, but I agree with you for future releases.

File: art/estimators/object_detection/python_object_detector.py (hunk 1 above)
Comment (beat-buesser) [reply art/estimators/object_detection/python_object_detector.py]: Good point. Following the previous item, we don't yet support tensors, but it's good to fix it now.
